projet/projects/views.py

```diff
@@ -250,4 +252,140 @@ def merge_features_documentation(request, feature_id):
+
+    def patch(self, request, project_id, fromat=None):
+        budget = Budget.objects.get(project__id=project_id)
+        serializer = BudgetSerializer(budget, data=request.data)
+        if request.user != budget.project.user:
+            error = "Only project managers/owners can update budget info."
+            return Response(error, status=status.HTTP_401_UNAUTHORIZED)
+        if serializer.is_valid():
+            from_amount = budget.amount
+            to_amount = request.data["amount"]
+            new_history = BudgetHistory.objects.create(
+                budget=budget, from_amount=from_amount, to_amount=to_amount
+            )
+            serializer.save()
+            return Response(serializer.data, status=status.HTTP_202_ACCEPTED)
+        error = "Unknown error"
+        return Response(error, status=status.HTTP_404_NOT_FOUND)
+
+
+@api_view(["GET", "DELETE"])
+def budget_history(request, project_id, history_id=None):
+    if request.method == "GET":
+        history = BudgetHistory.objects.filter(budget__project__id=project_id).order_by(
+            "-date_updated"
+        )
+        serializer = BudgetHistorySerializer(history, many=True)
+        return Response(serializer.data, status=status.HTTP_200_OK)
+    if request.method == "DELETE":
+        if history_id:
+            try:
+                history = BudgetHistory.objects.get(
+                    budget__project__id=project_id, id=history_id
+                )
+                history.delete()
+                message = "History deleted."
+                return Response(message, status=status.HTTP_200_OK)
+            except BudgetHistory.DoesNotExist:
+                error = "History not found."
+                return Response(error, status=status.HTTP_404_NOT_FOUND)
+    return Response("Unknown error.", status=status.HTTP_302_FOUND)
+
+
+class CreateExpense(APIView):
+    def post(self, request, project_id=None, format=None):
+        try:
+            project = Project.objects.get(id=project_id)
+        except Project.DoesNotExist:
+            error = "Project does not exist."
+            return Response(error, status=status.HTTP_404_NOT_FOUND)
+        test = TestIfContributor(user=request.user, project=project)
+
+        if test.test_is_contributor_or_manager() is not True:
+            error = "You have to be an accepted contributor in order to work on this project."
+            return Response(error, status=status.HTTP_401_UNAUTHORIZED)
+
+        data = request.data
+        data["contributor"] = request.user.id
+        serializer = ExpenseSerializer(data=data)
+        if serializer.is_valid():
+            serializer.save(project=project)
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        else:
+            error = "Unknown error"
+            return Response(error, status=status.HTTP_404_NOT_FOUND)
+
+
+@api_view(["GET", "PATCH", "DELETE"])
+@permission_classes([IsContributor])
```

Comment:
Expense detail, update, and delete view.
Only creators of and expense and project managers can delete or update them.